### PR TITLE
2FA (Branch)

### LINKED ISSUE #13: Implement 2FA to for user login
Show qr code and store user hash in db, valide code on login

### PR DESCRIPTION
**Pull request by hp355837**
Created a system to enable and disable 2fa for a user. User logs in. Goes to settings and there's a new page to enable 2fa. To enable it user has to generate an image then scan it and enter the one time password he gets. After it's enabled the user can check the password again if he wishes or disable it. If it's enabled then on the next login the user is asked to enter a one time password to login successfully.

@romilrobtsenkov @raimop

closes #13

### CODE REVIEW

File: local-dev/docker-compose.yml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 version: '3.4'
```
Comment (romilrobtsenkov) [inline local-dev/docker-compose.yml:1]: no changes should be done within this PR to this file

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (romilrobtsenkov) [inline .vscode/settings.json:1]: unneeded changes I guess?

File: teemaderegister-be/src/controllers/admin.js [hunk 3]
```diff
@@ -1,46 +1,5 @@
-const User = require('../models/user')
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/admin.js:1]: ? why this is removed

File: teemaderegister-be/src/controllers/users.js [hunk 4]
```diff
@@ -59,11 +67,14 @@ module.exports.getProfile = async (req, res) => {
...
     .select(`
+      
       -login.password 
       -login.passwordResetToken 
       -login.passwordResetExpires 
       -login.passwordUpdatedAt 
-      -profile.image.original
+      -user.profile.image.original
+
+
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/users.js:77]: extra lines, check other files also

File: teemaderegister-be/src/models/user.js [hunk 5]
```diff
@@ -73,8 +78,11 @@ userSchema.pre('save', async function (next) {
 * Helper method for validating user's password on login through user.comparePassword
 * Important! Do not use arrow function, will lose ref to (this)
 */
-userSchema.methods.comparePassword = function (candidatePassword) {
-  return bcrypt.compare(candidatePassword, this.login.password)
+userSchema.methods.comparePassword = async function (candidatePassword) {
+  const value = await bcrypt.compare(candidatePassword, this.login.password)
+  console.log(this.login.password)
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/models/user.js:83]: extra console logs, check other files also

File: teemaderegister-be/src/routes.js [hunk 6]
```diff
@@ -41,12 +41,16 @@ router.get('/users/profile', jwtEnsure, asyncMiddleware(users.getProfile))
...
+router.get('/factor', jwtEnsure, asyncMiddleware(factor.get))
+router.post('/factor/insert', jwtEnsure, asyncMiddleware(factor.insert))
+
 router.put('/users/reset-picture', jwtEnsure, asyncMiddleware(users.resetPicture))
 
 // SAMPLE
 router.get('/admin/', jwtEnsure, allowRoles([ADMIN]), asyncMiddleware(admin.getSecret))
-router.post('/admin/createUser', jwtEnsure, allowRoles([ADMIN]), asyncMiddleware(admin.createUser))
-router.post('/admin/tos/save', jwtEnsure, allowRoles([ADMIN]), asyncMiddleware(tos.saveTos))
-router.get('/tos', asyncMiddleware(tos.getTos))
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/routes.js:50]: ? why is this removed?

File: teemaderegister-be/src/utils/mail.js [hunk 7]
```diff
@@ -45,8 +44,6 @@ module.exports.sendMail = (options) => {
     text
   }
 
-  log.debug(mailOptions)
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/utils/mail.js:48]: these should stay here

**Review by romilrobtsenkov — COMMENTED** [review]
Screenshot 2021-01-11 at 09 43 38

would place this button better, also if possible would trigger the action automatically without that button press

**Comment by romilrobtsenkov** [comment]
Uploading Screenshot 202
![Screenshot 2021-01-11 at 09 45 59
1-01-11 at 09.46.13.png…]()

this is little bit confusing, would change the button labels if 2FA is enabled, also maybe there is no need to show check button then

**Comment by hp355837** [comment]
@romilrobtsenkov  @raimop Added back the removed functions and problems caused by eslint. Rest of the files are managed by eslint styles. Nothing should be missing now. Checker is removed and more labels for clarification are added.

File: .vscode/settings.json (hunk 2 above)
Comment (romilrobtsenkov) [reply .vscode/settings.json:1]: Would still remove this file

File: teemaderegister-be/src/controllers/auth.js [hunk 8]
```diff
@@ -25,9 +25,13 @@ module.exports.localLogin = async (req, res) => {
   if (!attemptAllowed) throw new Error('Too many unsuccessful login attempts. Check your email.')
 
   const isMatch = await user.comparePassword(password)
+  console.log(JSON.stringify(isMatch))
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/auth.js:28]: unneeded log

File: teemaderegister-be/src/controllers/auth.js [hunk 9]
```diff
@@ -25,9 +25,13 @@ module.exports.localLogin = async (req, res) => {
...
   if (!isMatch) {
     throw new Error('Email or password incorrect')
   }
+  if (user.auth.enabled) {
+    if (!token) throw new InsertToken('Please insert token!')
+  }
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/auth.js:34]: maybe not the best way but currently could not think of better way

File: teemaderegister-be/src/controllers/auth.js [hunk 10]
```diff
@@ -90,7 +94,9 @@ module.exports.forgotPassword = async (req, res) => {
   }
   user.login = {
     ...user.login,
-    ...this.getPasswordResetTokenValues()
+    ...this.getPasswordResetTokenValues(),
+    passwordResetToken: crypto.randomBytes(20).toString('hex'),
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/auth.js:98]: those lines 98 and 99 seem unneeded changes

File: teemaderegister-be/src/controllers/factor.js [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+const qrcode = require('qrcode')
+
+const secret = speakeasy.generateSecret({
+  name: process.env.QR_SECRET
+})
+
+module.exports.create = async (req, res) => {
+  const {_id} = req.user
+  const user = await User.findOne({_id})
+  try {
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/factor.js:12]: no need to use try-catch

File: teemaderegister-be/src/controllers/factor.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    if (user.factor) return res.json('Cannot add anymore!')
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/controllers/factor.js:13]: would return response in error form

File: teemaderegister-be/src/controllers/users.js (hunk 4 above)
Comment (romilrobtsenkov) [reply teemaderegister-be/src/controllers/users.js:77]: still valid

File: teemaderegister-be/src/models/user.js [hunk 13]
```diff
@@ -73,8 +78,9 @@ userSchema.pre('save', async function (next) {
 * Helper method for validating user's password on login through user.comparePassword
 * Important! Do not use arrow function, will lose ref to (this)
 */
-userSchema.methods.comparePassword = function (candidatePassword) {
-  return bcrypt.compare(candidatePassword, this.login.password)
+userSchema.methods.comparePassword = async function (candidatePassword) {
+  const value = await bcrypt.compare(candidatePassword, this.login.password)
+  return value
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/models/user.js:83]: this function can still return a promise to be used in await format, but keep it as is, its ok

File: teemaderegister-be/src/routes.js [hunk 14]
```diff
@@ -41,12 +42,18 @@ router.get('/users/profile', jwtEnsure, asyncMiddleware(users.getProfile))
 router.put('/users/profile', jwtEnsure, validate.userAccountUpdate, asyncMiddleware(users.updateUser))
 router.put('/users/password', jwtEnsure, validate.userPasswordUpdate, asyncMiddleware(users.updatePassword))
 router.post('/users/upload-picture', jwtEnsure, multerMiddleware('profileImage'), asyncMiddleware(users.uploadPicture))
+
+router.post('/factor', jwtEnsure, asyncMiddleware(factor.create))
```
Comment (romilrobtsenkov) [inline teemaderegister-be/src/routes.js:46]: would prefix them still with auth/local... and then add /factor. As they are for local auth

File: teemaderegister-react/src/actions/CurriculumActions.js [hunk 15]
```diff
@@ -30,6 +30,7 @@ export const getCurriculums = () => dispatch => {
     .then(data => dispatch({
       type: types.CURRICULUMS_LOADED,
       curriculums: data.curriculums
+
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/actions/CurriculumActions.js:33]: extra line not needed

File: teemaderegister-react/src/components/Settings2fa.jsx [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+                <Form
+                  name='basic'
+                  onFinish={onClose}
+                >
+                  <h2 >Currently the 2FA is Enabled</h2>
+                  <h3 >Insert the QR token and press Disable to deactivate 2fa</h3>
+                  <Form.Item
+                    label='Token'
+                    name='token'
+                    style={{ width: 200, margin: '0 10px' }}
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/components/Settings2fa.jsx:72]: rather add style via classname and CSS, it is not recommended to user style attribute

File: teemaderegister-react/src/index.jsx [hunk 17]
```diff
@@ -108,8 +113,9 @@ render(
           </Content>
           <Footer className='layout__footer'>
             <a href={links.project}> Teemaderegister</a><br/>
+            <a href={links.project}><GithubOutlined /> Teemaderegister</a><br/>
             Code licensed under <a href={links.license}>MIT License</a><br/>
-            Content © 2010-{new Date().getFullYear()} <a href={links.content}>Tallinn University</a><br/>
+            Content © 2010-{new Date().getFullYear()} <a href={links.content}>Tallinn University</a>
             <Link to={TOS_PATH}>Terms of Service</Link>
```
Comment (romilrobtsenkov) [inline teemaderegister-react/src/index.jsx:119]: not sure if those changes are intended

File: teemaderegister-react/src/components/Settings2fa.jsx (hunk 16 above)
Comment (hp355837) [reply teemaderegister-react/src/components/Settings2fa.jsx]: Done and moved to css

File: teemaderegister-be/src/routes.js (hunk 14 above)
Comment (hp355837) [reply teemaderegister-be/src/routes.js:46]: Done setup now  with auth/local

File: teemaderegister-be/src/controllers/factor.js (hunk 11 above)
Comment (hp355837) [reply teemaderegister-be/src/controllers/factor.js:12]: Removed the try catch

File: teemaderegister-react/src/index.jsx (hunk 17 above)
Comment (hp355837) [reply teemaderegister-react/src/index.jsx:119]: Possible that Eslint did it not me

**Review by romilrobtsenkov — APPROVED** [review]
